### ISSUE TITLE
Failure to add new applications to running cluster with aad-pod-identity

### ISSUE THREAD

**Issue by galvesribeiro** [Azure/aad-pod-identity#730]
Hello folks!

I've tried to add a new application to our existing AKS cluster with aad-pod-identity and it looks like the application is not able to acquire a token using the ManagedIdentityCredential anymore.

The application POD logs this:

```
pl-identity
Unhandled exception. Azure.Identity.AuthenticationFailedException: ManagedIdentityCredential authentication failed.
pl-identity
---> Azure.RequestFailedException: Service request failed.
pl-identity
Status: 500 (Internal Server Error)
pl-identity
Content:
pl-identity
match failed, ip:100.64.1.68 matching pods:[]
pl-identity
Headers:
pl-identity
X-Content-Type-Options: REDACTED
pl-identity
Date: Sat, 08 Aug 2020 22:24:52 GMT
pl-identity
Content-Type: text/plain; charset=utf-8
pl-identity
Content-Length: 46
pl-identity
at Azure.Identity.ManagedIdentityClient.AuthenticateAsync(String[] scopes, CancellationToken cancellationToken)
pl-identity
at Azure.Identity.ManagedIdentityCredential.GetTokenImplAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
pl-identity
--- End of inner exception stack trace ---
pl-identity
at Azure.Identity.ExtendedAccessToken.GetTokenOrThrow()
pl-identity
at Azure.Identity.ManagedIdentityCredential.GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken)
pl-identity
at Azure.Core.Pipeline.BearerTokenAuthenticationPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
pl-identity
at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
pl-identity
at Azure.Core.Pipeline.RetryPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline, Boolean async)
pl-identity
at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline)
pl-identity
at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline)
pl-identity
at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline)
pl-identity
at Azure.Core.Pipeline.HttpPipelineSynchronousPolicy.ProcessAsync(HttpMessage message, ReadOnlyMemory`1 pipeline)
pl-identity
at Azure.Core.Pipeline.HttpPipeline.SendRequestAsync(Request request, CancellationToken cancellationToken)
pl-identity
at Azure.Data.AppConfiguration.ConfigurationClient.GetConfigurationSettingsPageAsync(SettingSelector selector, String pageLink, CancellationToken cancellationToken)
pl-identity
at Azure.Core.PageResponseEnumerator.FuncAsyncPageable`1.AsPages(String continuationToken, Nullable`1 pageSizeHint)+MoveNext()
pl-identity
at Azure.Core.PageResponseEnumerator.FuncAsyncPageable`1.AsPages(String continuationToken, Nullable`1 pageSizeHint)+System.Threading.Tasks.Sources.IValueTaskSource.GetResult()
pl-identity
at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()
pl-identity
at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+MoveNext()
pl-identity
at Azure.AsyncPageable`1.GetAsyncEnumerator(CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource.GetResult()
pl-identity
at Microsoft.Extensions.Configuration.AzureAppConfiguration.AzureAppConfigurationProvider.<>c__DisplayClass15_3.<b__4>d.MoveNext()
pl-identity
--- End of stack trace from previous location where exception was thrown ---
pl-identity
at Microsoft.Extensions.Configuration.AzureAppConfiguration.AzureAppConfigurationProvider.<>c__DisplayClass15_3.<b__4>d.MoveNext()
pl-identity
--- End of stack trace from previous location where exception was thrown ---
pl-identity
at Microsoft.Extensions.Configuration.AzureAppConfiguration.TracingUtils.CallWithRequestTracing(Boolean tracingEnabled, RequestType requestType, HostType hostType, Func`1 clientCall)
pl-identity
at Microsoft.Extensions.Configuration.AzureAppConfiguration.AzureAppConfigurationProvider.CallWithRequestTracing(Func`1 clientCall)
pl-identity
at Microsoft.Extensions.Configuration.AzureAppConfiguration.AzureAppConfigurationProvider.LoadAll(Boolean ignoreFailures)
pl-identity
at Microsoft.Extensions.Configuration.AzureAppConfiguration.AzureAppConfigurationProvider.Load()
pl-identity
at Microsoft.Extensions.Configuration.ConfigurationRoot..ctor(IList`1 providers)
pl-identity
at Microsoft.Extensions.Configuration.ConfigurationBuilder.Build()
pl-identity
at Microsoft.Extensions.Hosting.HostBuilder.BuildAppConfiguration()
pl-identity
at Microsoft.Extensions.Hosting.HostBuilder.Build()
pl-identity
at PL.Identity.Program.Main(String[] args) in /Users/guto/dev/repos/PagoLivre/Identity/src/PL.Identity/Program.cs:line 14
```

Both NMI pods are reporting lots of this:

```
nmi
E0808 22:21:48.330401 1 client.go:184] pod list empty
nmi
W0808 22:21:48.330430 1 client.go:210] List pod error: . Retrying, attempt number: 1
nmi
E0808 22:21:48.630644 1 client.go:184] pod list empty
nmi
W0808 22:21:48.630696 1 client.go:210] List pod error: . Retrying, attempt number: 2
nmi
E0808 22:21:48.930860 1 client.go:184] pod list empty
nmi
W0808 22:21:48.930911 1 client.go:210] List pod error: . Retrying, attempt number: 3
nmi
E0808 22:21:49.231058 1 client.go:184] pod list empty
nmi
W0808 22:21:49.231097 1 client.go:210] List pod error: . Retrying, attempt number: 4
nmi
E0808 22:21:49.531240 1 client.go:184] pod list empty
nmi
E0808 22:21:49.531307 1 server.go:342] missing podname for podip:100.64.1.68, match failed, ip:100.64.1.68 matching pods:[]
nmi
I0808 22:21:49.531322 1 server.go:191] Status (500) took 1200940401 ns for req.method=GET reg.path=/metadata/identity/oauth2/token req.remote=100.64.1.68
```

And the MIC logs has this:

```
mic
I0808 22:12:08.018194 1 stats.go:129] *********************
mic
I0808 22:18:19.243215 1 mic.go:1009] Processing node aks-plpool1-XXXXXXXXXXXXX-vmss, add [1], del [0], update [0]
mic
I0808 22:18:19.243249 1 crd.go:515] Got assigned id pl-identity-dev-69b55bc7bc-9btp4-dev-pl-identity-dev to assign
mic
I0808 22:18:19.463036 1 cloudprovider.go:208] Updating user assigned MSIs on aks-plpool1-XXXXXXXXXXXXX-vmss, assign [1], unassign [0]
mic
I0808 22:18:59.614412 1 crd.go:793] Updating assigned identity default/pl-identity-dev-69b55bc7bc-9btp4-dev-pl-identity-dev status to Assigned
mic
I0808 22:18:59.622363 1 mic.go:514] Work done: true. Found 9 pods, 9 ids, 9 bindings
mic
I0808 22:18:59.622374 1 mic.go:515] Total work cycles: 882, out of which work was done in: 17
mic
I0808 22:18:59.622385 1 stats.go:99] ** Stats collected **
mic
I0808 22:18:59.622390 1 stats.go:82] Pod listing: 28.301µs
mic
I0808 22:18:59.622399 1 stats.go:82] ID listing: 13.801µs
mic
I0808 22:18:59.622402 1 stats.go:82] Binding listing: 11.6µs
mic
I0808 22:18:59.622405 1 stats.go:82] Assigned ID listing: 39.601µs
mic
I0808 22:18:59.622408 1 stats.go:82] System: 269.007µs
mic
I0808 22:18:59.622410 1 stats.go:82] CacheSync: 2.1µs
mic
I0808 22:18:59.622414 1 stats.go:82] Cloud provider get: 195.743149ms
mic
I0808 22:18:59.622424 1 stats.go:82] Cloud provider put: 40.151304505s
mic
I0808 22:18:59.622427 1 stats.go:82] Assigned ID addition: 23.977718ms
mic
I0808 22:18:59.622430 1 stats.go:82] Assigned ID deletion: 0s
mic
I0808 22:18:59.622433 1 stats.go:89] Number of cloud provider PUT: 1
mic
I0808 22:18:59.622435 1 stats.go:89] Number of cloud provider GET: 1
mic
I0808 22:18:59.622438 1 stats.go:89] Number of assigned ids created in this sync cycle: 1
mic
I0808 22:18:59.622440 1 stats.go:89] Number of assigned ids updated in this sync cycle: 0
mic
I0808 22:18:59.622443 1 stats.go:89] Number of assigned ids deleted in this sync cycle: 0
mic
I0808 22:18:59.622445 1 stats.go:82] Find assigned ids to create: 9.8µs
mic
I0808 22:18:59.622447 1 stats.go:82] Find assigned ids to delete: 6.7µs
mic
I0808 22:18:59.622450 1 stats.go:82] Total time to assign or remove IDs: 40.379138983s
mic
I0808 22:18:59.622452 1 stats.go:82] Event recording: 0s
mic
I0808 22:18:59.622455 1 stats.go:82] Total: 40.379556993s
mic
I0808 22:18:59.622457 1 stats.go:129] *********************
```

The cluster used to be one node only before we went to production and now it has 2 or more, made the aad-pod-identity pods to be spread among the nodes as expected.

Does anyone know what are we missing here?

Thanks!

**Comment by galvesribeiro** [comment]
Just to rule out that the application has wrong manifests, I've tried to deploy the demo pod from the README and here is its output:

```
demo
time="2020-08-09T21:24:10Z" level=error msg="failed to refresh ServicePrincipalTokenFromMSI using the MSI VM extension, msiEndpoint(http://169.254.169.254/metadata/identity/oauth2/token)" podip=100.64.1.59 podname=demo podnamespace=demo
demo
time="2020-08-09T21:24:10Z" level=error msg="testMSIEndpoint failed, " podip=100.64.1.59 podname=demo podnamespace=demo
demo
time="2020-08-09T21:24:21Z" level=error msg="failed list all vm azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions//resourceGroups//providers/Microsoft.Compute/virtualMachines?api-version=2017-12-01: StatusCode=500 -- Original Error: adal: Refresh request failed. Status Code = '500'. Response body: match failed, ip:100.64.1.59 matching pods:[]\n" podip=100.64.1.59 podname=demo podnamespace=demo
demo
time="2020-08-09T21:24:32Z" level=error msg="failed to refresh ServicePrincipalTokenFromMSI using the MSI VM extension, msiEndpoint(http://169.254.169.254/metadata/identity/oauth2/token)" podip=100.64.1.59 podname=demo podnamespace=demo
demo
time="2020-08-09T21:24:32Z" level=error msg="testMSIEndpoint failed, " podip=100.64.1.59 podname=demo podnamespace=demo
```

**Comment by chewong** [comment]
Could you share the AzureIdentity and AzureIdentityBinding that you deployed? Are you also using namespaced mode?

**Comment by galvesribeiro** [comment]
The identity objects are generated from this helm templates:

```yaml
apiVersion: "aadpodidentity.k8s.io/v1"
kind: AzureIdentity
metadata:
  name: {{ .Values.identity.name }}
spec:
  type: 0
  resourceID: {{ .Values.identity.resourceId }}
  clientID: {{ .Values.identity.clientId }}

---

apiVersion: "aadpodidentity.k8s.io/v1"
kind: AzureIdentityBinding
metadata:
  name: "{{ .Values.identity.name }}-binding"
spec:
  azureIdentity: {{ .Values.identity.name }}
  selector: {{ .Values.identity.name }}
```

We are monitoring all namespaces. It works across all the namespaces we have today, and the issue also happens across all of them.

**Comment by chewong** [comment]
Based on your MIC logs, it looks like there is an error when NMI is listing pods - `List pod error: . Retrying, attempt number: 1` but the error wasn't being logged correctly (will open a PR to address that). 

Are you using 1.6.2? If not, could you give it a try and see if upgrading aad-pod-identity would resolve the issue?

**Comment by galvesribeiro** [comment]
We are on 1.6.2. As I mentioned on #729 the only way to get it working was to remove the managed identity, force upgrade the chart, and then redeploy the app using a new managed identity with a new name.

**Comment by galvesribeiro** [comment]
@aramase being back here from your last commend [here](https://github.com/Azure/aad-pod-identity/issues/725#issuecomment-672329075), I've noticed another situation that cause this weird behaviour.

I've just deployed a new application using Helm, which starts with a deployment and 1 pod and a freshy, just created new Azure Identity. It fail with the same error however, if I kill both NMI pods, wait it to restart, and then kill the pod failing from the application, the new one that comes up will work just fine. It doesn't stop there... If I edit the Deployment and set 2 replicas, the new pod will suffer the same problem when trying to acquirer a token:

```
pl-identity
Unhandled exception. Azure.Identity.AuthenticationFailedException: ManagedIdentityCredential authentication failed.
pl-identity
---> Azure.RequestFailedException: Service request failed.
pl-identity
Status: 500 (Internal Server Error)
pl-identity
Content:
pl-identity
failed to match pod IP 100.64.1.85 with []
```
Then if I go again and restart NMI pods, kill the new app pod, and let the Deployment create it again, it works.

I think there is something that is happening when we are behind Azure Firewall that is blocking the NMI or whatever it is doing but I can't find out from the logs... That could be preventing NMI to somehow talk to the Kube API and get the pod names based on its IP as you mentioned.

If you want to discuss offline, we can do a screen share session and I can show you that happening.
